pkg/app/appevent/subscriber.go

```diff
@@ -0,0 +1,123 @@
+package appevent
+
+import (
+	"errors"
+	"sync"
+)
+
+// subChanSize is used so that incoming events are kept in order
+const subChanSize = 5
+
+// Errors associated with the Subscriber type.
+var (
+	ErrSubscriptionsClosed = errors.New("event subscriptions is closed")
+)
+
+// Subscriber is used by apps and contain subscription channels to different event types.
+type Subscriber struct {
+	chanSize int // config: event channel size
+
+	m      map[string]chan *Event
+	mx     sync.RWMutex
+	closed bool
+}
+
+// NewSubscriber returns a new Subscriber struct.
+func NewSubscriber() *Subscriber {
+	return &Subscriber{
+		chanSize: subChanSize,
+		m:        make(map[string]chan *Event),
+		closed:   false,
+	}
+}
+
+// TCPDial subscribes to the TCPDial event channel.
+// This should only be called once.
+func (s *Subscriber) TCPDial(action func(data TCPDialData)) {
+	s.mx.Lock()
+	ch := make(chan *Event, s.chanSize)
```

Comment:
Can't we just check if the event already presents in map? And if it is, simply return. This way we may allow to call this func more than once safely

Reply:
This one is true for all subscribing funcs

Reply:
@Darkren But what is the usecase for calling subscribing funcs multiple times? Should we end up triggering all the `action` input funcs of all `.TCPDial` call?